### PR TITLE
serial:update some bug fix,and functional improvements

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
update some bug fix,and functional improvements

Fix system apps cu_main() ctrl + c crash bug:
1. fix cu ctrl+c crash.
2. use nxsig_tgkill instead of nxsig_kill

Fix syslog can't output after serial setup:
3. serial/pl011: fix can't up_putc() after setup

Fix poll crash:
4. serial.c: should protect the fds when calling poll_notify

Fix PCI serial driver init crash:
5. make sure that interrupts are disabled during init
This is a proper fix for an unexpected MSI interrupt for PCI serial driver.

Feature enhancements --- force panic feature use two ctrl+c
6. support force panic only when repeat the panic char

Feature enhancements --- Add new option
7.  Add 16550_DLF_SIZE option for DesignWare UART.

## Impact
serial
## Testing
vela

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



## NuttX PR Requirements Review

While this PR description provides a decent overview of the changes made, it lacks the specificity required by the NuttX PR guidelines. 

Here's a breakdown of what needs improvement:

**Summary:**

* **Why is each change necessary?** Briefly explain the reasoning behind each bug fix and the motivation for each functional improvement.
* **What functional part of the code is being changed?**  Be more specific than "serial". Mention the affected modules and files.
* **How does the change exactly work?** Provide a concise explanation of the technical implementation for each change.

**Impact:**

* **Is new feature added?**  Clearly state "YES" and list the new features.
* **Is existing feature changed?** State "YES" and describe how existing features are modified.
* **Impact on user:**  Will any user-facing behavior change? If yes, describe how.
* **Impact on build:** Does the PR introduce any new build options or dependencies? 
* **Impact on hardware:** Specify the affected architectures, boards, and drivers.
* **Impact on documentation:** Does this PR require documentation updates? If yes, are they included?
* **Impact on security:** Analyze if the changes have any security implications (positive or negative).
* **Impact on compatibility:**  Address potential backward/forward compatibility concerns. 

**Testing:**

* **Build Host(s):** Specify the operating system, CPU architecture, and compiler used for testing.
* **Target(s):** Be specific about the target architecture, board, and configuration.
* **Testing logs:** Provide relevant snippets of testing logs before and after the changes.

**Overall:**

The PR description is a good starting point but needs to be more detailed and specific to meet the NuttX requirements. Ensure to address each point in the guidelines with clear and concise information.

**Review by jerpelea — CHANGES_REQUESTED** [review]
please add a commit message with explanation for this change

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by nuttxpr** [comment]
Sorry @anjiahao1: This commit seems to hang the Console Output on Milk-V Duo S (based on 16550 UART):

```text
Bytes transferred = 21575 (5447 hex)
## Flattened Device Tree blob at 81200000
   Booting using the fdt blob at 0x81200000
   Loading Device Tree to 000000009f275000, end 000000009f27d446 ... OK
Starting kernel 
```

It works OK with the previous commit: [c00d4776711](https://github.com/apache/nuttx/tree/c00d47767118785eec1aab2433ec87229af6a489). Any idea why this causes problems on Milk-V Duo S? Thanks!

**Comment by lupyuen** [comment]
Oops that was me posting the above messages. Please lemme know what you think, thanks for looking into this :-)

File: drivers/serial/serial.c [hunk 1]
```diff
@@ -154,6 +159,28 @@ static struct work_s g_serial_work;
...
+ ****************************************************************************/
+
+static void uart_poll_notify(FAR uart_dev_t *dev, unsigned int min,
+                             unsigned int max, pollevent_t eventset)
+{
+  irqstate_t flags;
+
+  DEBUGASSERT(max > min && max - min <= CONFIG_SERIAL_NPOLLWAITERS);
+
+  flags = enter_critical_section();
```
Comment (jlaitine) [inline drivers/serial/serial.c:173]: I don't think that this fix is correct;

The race condition is inside poll_notify(), when setting the revents bits:
https://github.com/apache/nuttx/blob/904b6ff85b479abe6c75ae2e3e1fda3af9e7a84f/fs/vfs/fs_poll.c#L289

While the poll_notify is being called both from irq and from task context, the read-modify-write needs to be protected there with a critical section in there. This fixes just a symptom for uart driver - the underlying issue is still there in the poll interface and affects other driver using the poll interface.

The sched lock should not be needed while the call to the callback in poll_notify is left out of the critical section - unless you need to protect that the fds->revents doesn't get new bits during a call to callback. But I don't really see the need for that ?

Comment (xiaoxiang781216) [reply drivers/serial/serial.c:173]: sure, let's continue discuss in your new pr.

**Pull request merged**